### PR TITLE
Extraction from MatterDevice and MatterController

### PR DESCRIPTION
**Pull request by lauckhart**
Most of this involves moving logic out of MatterDevice into topical components.  Modifies MatterController along the same lines though mostly focused on logic that could be shared with MatterDevice.

New components include:

  - DeviceCommissioner - server-side equivalent to ControllerCommissioner
  - DeviceAdvertiser - probably deserves additional cleanup but handles MDNS and BLE announcement for devices
  - NodeFinder - currently focused on operational discover but could also add commissioning discovery
  - Subscription - base class for client/server subscriptions (including ServerSubscription, previously SubscriptionHandler)
  - TransportInterfaceSet/NetInterfaceSet - collections of network connections with a handful of utility methods

Existing components also receive a variety of changes.  Mostly to make them interface directly without requiring a than relying on a higher-level coordinater like MatterDevice.

Used a few code patterns to support this:

  - Replaced more callbacks with events.  This allows interested parties to subscribe independently
  - In several places eventing comes via a mutable/observable set, so e.g. "session.subscriptions.added"
  - Many of the components now take a "context" as input that includes component dependencies
  - The classes generally include an (optional) static method that extracts components from an Environment
  - Moved async initialization from some classes from an "initialize" method into a construction field

There should be minimal breaking changes at the MatterDevice/MatterController level.  Lower levels do have breaking changes.

Also includes early development of controller/binding implementation for Node API.  This is related to above but including only because developed in tandem and doesn't affect existing API.

### CODE REVIEW

File: packages/protocol/src/protocol/NodeFinder.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+     * The scanners this NodeFinder supports.
+     */
+    get scanners() {
+        return this.#scanners;
+    }
+
+    /**
+     * Find a node for operational (non-commissioning) purposes.
+     */
+    async connectToPeer(fabric: Fabric, nodeId: NodeId, timeoutSeconds = 5): Promise<undefined | PeerConnection> {
```
Comment (Apollon77) [inline packages/protocol/src/protocol/NodeFinder.ts:73]: for completeness ... I think this is curremtly not used at all ... so could be replaced by the stuff in MatterController class

File: packages/protocol/src/session/InsecureSession.ts [hunk 2]
```diff
@@ -95,6 +90,7 @@ export class InsecureSession extends Session {
     }
 
     async end() {
-        await this.closeCallback?.();
+        logger.info(`End insecure session ${this.name}`);
+        this.manager?.insecureSessions.delete(this.nodeId);
```
Comment (Apollon77) [inline packages/protocol/src/session/InsecureSession.ts:94]: really such an access from "outside"? session could also emit an event when closed and sessionmanager reacts and deletes? Same on SecureSession do we really want to wire the sessionmanager into the session? That somehow feels wrong

File: packages/tools/src/building/typescript.ts [hunk 3]
```diff
@@ -116,7 +116,7 @@ export class Typescript {
             host: this.#host,
         });
 
-        // See https://github.com/microsoft/TypeScript/issues/31849
+        // See https://github.com/microsoft/TypeScript/issues/31849p
```
Comment (Apollon77) [inline packages/tools/src/building/typescript.ts:119]:
```suggestion
        // See https://github.com/microsoft/TypeScript/issues/31849
```

**Review by Apollon77 — COMMENTED** [review]
ok first round of checks ...open is still all the logic moved out of Matterdevice in their new places

File: packages/protocol/src/protocol/DeviceCommissioner.ts [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        // Remove PASE responder when we close enhanced commissioning window or node is commissioned
+        if (
+            this.#windowStatus === AdministratorCommissioning.CommissioningWindowStatus.EnhancedWindowOpen ||
+            this.#context.fabrics.getFabrics().length > 0
+        ) {
+            this.#context.secureChannelProtocol.removePaseCommissioner();
+        }
+        this.#windowStatus = AdministratorCommissioning.CommissioningWindowStatus.WindowNotOpen;
+
+        await this.#context.advertiser.exitCommissioningMode();
```
Comment (Apollon77) [inline packages/protocol/src/protocol/DeviceCommissioner.ts:261]: I thuink if you move this line below the callback handling block the error is gone ... the question is still why this consistently fails/blocks, but the "old" code has these two blocks exchanged

File: CHANGELOG.md [hunk 5]
```diff
@@ -21,6 +21,8 @@ The main work (all changes without a GitHub username in brackets in the below li
     -   Info: General functionality that is not Matter specific previously resided in `@project-chip/matter.js`.  It now lives in `@matter.js/general`
     -   BREAKING: The "ByteArray" type is removed, replaced with native-JS Uint8Array and a small collection of utility functions in the "Bytes" namespace
     -   Feature: The default "Time" implementation is now fully functional across all standard JS runtimes
+    -   Enhancement: Network transports can now self select which the protocols and addresses they support
+    -   Feature: A new `ObserverGroup` class simplifies binding binding management for multiple observables
```
[the remark this answers is not in the document]
Comment (lauckhart) [reply CHANGELOG.md:25]: done

File: packages/general/src/util/Observable.ts [hunk 6]
```diff
@@ -461,3 +461,88 @@ export class ObservableProxy extends BasicObservable {
...
+                return;
+            }
+            observers.delete(target);
+            if (observers.size === 0) {
+                this.#boundObservers.delete(observer);
+            }
+        }
+        const observers = this.#observers.get(observable);
+        const index = observers?.indexOf(observer);
+        if (index !== undefined) {
```
[the remark this answers is not in the document]
Comment (lauckhart) [reply packages/general/src/util/Observable.ts:530]: Right that's a bug, fixed.  This code isn't actually used anywhere, just added for completeness.

File: packages/matter.js/src/CommissioningServer.ts [hunk 7]
```diff
@@ -587,13 +581,28 @@ export class CommissioningServer extends MatterNode {
                 }
             },
             (fabricIndex: FabricIndex) => this.options.activeSessionsChangedCallback?.(fabricIndex),
-            { maxPathsPerInvoke: this.interactionServer.maxPathsPerInvoke },
+            { maxPathsPerInvoke: maxPathsPerInvoke },
```
[the remark this answers is not in the document]
Comment (lauckhart) [reply packages/matter.js/src/CommissioningServer.ts:584]: Right now I think it is, InteractionServer sets it from the same value passed in below and doesn't change it.

File: packages/matter.js/src/CommissioningServer.ts [hunk 8]
```diff
@@ -587,13 +581,28 @@ export class CommissioningServer extends MatterNode {
...
+            sessions: deviceInstance.sessionManager,
+            structure: this.endpointStructure,
+            subscriptionOptions: {
+                maxIntervalSeconds: this.options.subscriptionMaxIntervalSeconds,
+                minIntervalSeconds: this.options.subscriptionMinIntervalSeconds,
+                randomizationWindowSeconds: this.options.subscriptionRandomizationWindowSeconds,
+            },
+            maxPathsPerInvoke,
+            initiateExchange: (fabric, nodeId, protocolId) => {
+                return this.deviceInstance!.initiateExchange(fabric, nodeId, protocolId);
```
[the remark this answers is not in the document]
Comment (lauckhart) [reply packages/matter.js/src/CommissioningServer.ts:601]: done

File: packages/protocol/src/interaction/InteractionServer.ts [hunk 9]
```diff
@@ -1053,17 +1042,38 @@ export class InteractionServer implements ProtocolHandler, InteractionRecipient
                     message,
                     this.#endpointStructure.getEndpoint(path.endpointId)!,
                 ),
+
+            initiateExchange: (fabric, nodeId, protocolId) =>
+                this.#context.initiateExchange(fabric, nodeId, protocolId),
+        };
+
+        // TODO: Interpret specs:
+        // The publisher SHALL compute an appropriate value for the MaxInterval field in the action. This SHALL respect the following constraint: MinIntervalFloor ≤ MaxInterval ≤ MAX(SUBSCRIPTION_MAX_INTERVAL_PUBLISHER_LIMIT=60mn, MaxIntervalCeiling)
```
[the remark this answers is not in the document]
Comment (lauckhart) [reply packages/protocol/src/interaction/InteractionServer.ts:1051]: done

File: packages/matter.js/src/MatterController.ts [hunk 10]
```diff
@@ -335,33 +307,37 @@ export class MatterController implements SessionContext {
...
+        this.scanners = scanners;
+        this.netInterfaces = netInterfaces;
         this.certificateManager = certificateManager;
         this.fabric = fabric;
         this.sessionClosedCallback = sessionClosedCallback;
         this.adminVendorId = adminVendorId;
 
-        this.sessionManager = new SessionManager(this, sessionStorage);
-        this.sessionManager.sessionClosed.on(async session => {
-            if (!session.closingAfterExchangeFinished) {
```
[the remark this answers is not in the document]
Comment (lauckhart) [reply packages/matter.js/src/MatterController.ts:348]: See line 133 in ExchangeManager.ts, it reacts to sessions.deleted

File: packages/protocol/src/protocol/NodeFinder.ts (hunk 1 above)
Comment (lauckhart) [reply packages/protocol/src/protocol/NodeFinder.ts:73]: Damnit, something made me think it was.  I stubbed this method out for now and removed related code from MatterDevice

File: packages/protocol/src/session/InsecureSession.ts (hunk 2 above)
Comment (lauckhart) [reply packages/protocol/src/session/InsecureSession.ts:94]: Right now this also replaces the "everything in the world" context object required by the CommissioningServer code branch.  Looks like we could add a "closed" promise to the session and have the manager listen to that though I'd maybe wait until CommissioningServer goes away.

File: packages/tools/src/building/typescript.ts (hunk 3 above)
Comment (lauckhart) [reply packages/tools/src/building/typescript.ts:119]: done

File: packages/protocol/src/protocol/DeviceCommissioner.ts (hunk 4 above)
Comment (lauckhart) [reply packages/protocol/src/protocol/DeviceCommissioner.ts:261]: Ah OK, good catch.  Will try it

File: packages/matter.js/test/cluster/ClusterServerTest.ts [hunk 11]
```diff
@@ -572,7 +572,7 @@ describe("ClusterServer structure", () => {
             expect((server.attributes as any).generatedCommandList.get()).deep.equal([]);
         });
 
-        it("Verify init/destroy is called on CLusterServe when defined", async () => {
+        it("Verify init/destroy is called on ClusterServe when defined", async () => {
```
[the remark this answers is not in the document]
Comment (lauckhart) [reply packages/matter.js/test/cluster/ClusterServerTest.ts:575]: done

File: packages/protocol/src/session/InsecureSession.ts (hunk 2 above)
Comment (lauckhart) [reply packages/protocol/src/session/InsecureSession.ts:94]: I did make the manager getter protected so this is just an implementation detail now.

File: packages/protocol/src/interaction/ServerSubscription.ts [hunk 12]
```diff
@@ -734,16 +763,16 @@ export class SubscriptionHandler {
...
         if (attributeListenerData === undefined) return; // Ignore changes to attributes that are not subscribed to
 
         const { attribute } = attributeListenerData;
         if (attribute instanceof FabricScopedAttributeServer) {
             // We cannot be sure what value we got for fabric filtered attributes (and from which fabric),
             // so get it again for this relevant fabric. This also makes sure that fabric sensitive fields are filtered
-            // TODO: Remove this once we remove the legacy API and go away from using AttributeServers in the background
-            return this.readAttribute(path, attribute, true).then(({ value }) => {
-                this.outstandingAttributeUpdates.set(attributePathToId(path), {
+            // TODO: Maybe add try/catch when we add ACL handling and ignore the update if we cannot get the value?
```
[the remark this answers is not in the document]
Comment (lauckhart) [reply packages/protocol/src/interaction/ServerSubscription.ts]: fixed

**Pull request merged**